### ISSUE TITLE
Configuring `dataSourceProperties` for external databases can have unexpected behaviour

### ISSUE THREAD

**Issue by dazraf** [corda/corda#4037]
# Issue with `node.config` when setting up `dataSourceProperties` for an external database

## Summary

* The config doesn't work quite as stated [here](https://stackoverflow.com/questions/50232016/how-can-the-corda-node-be-extended-to-work-with-databases-other-than-h2/50232017#50232017) because of the way the config is presented to `HikariPool` from `AbstractNode`.
* There is a workaround but it would be good to either fix the configuration logic or to improve documentation.
* The `schema` field is ignored - there doesn't appear to be a way of setting which schema corda should be using and it defaults to the choice of the JDBC driver.

## Analysis

The following uses the Postgres driver. To set this up, with this configuration, the driver jar should be placed in `/Library/Java/postgres`.

Given this config in `deployNodes`:

```
        extraConfig = [
                jarDirs               : ['/Library/Java/postgres'],
                'dataSourceProperties': [
                        'dataSourceClassName': 'org.postgresql.ds.PGSimpleDataSource',
                        '"dataSource.url"'     : 'jdbc:postgresql://localhost:5432/postgres',
                        '"dataSource.user"'    : 'corda',
                        '"dataSource.password"': 'corda'
                ],
                'database'            : [
                        'transactionIsolationLevel': 'READ_COMMITTED',
                        'schema'                   : 'corda'
                ]
        ]
```

... `deployNodes` fails. The log shows:

```
[ERROR] 2018-10-04T11:16:05,466Z [main] util.PropertyElf.setProperty - Failed to set property url on target class org.postgresql.ds.PGSimpleDataSource {}
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_181]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_181]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_181]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_181]
        at com.zaxxer.hikari.util.PropertyElf.setProperty(PropertyElf.java:146) ~[HikariCP-2.5.1.jar:?]
        at com.zaxxer.hikari.util.PropertyElf.lambda$setTargetFromProperties$0(PropertyElf.java:57) ~[HikariCP-2.5.1.jar:?]
        at java.util.Hashtable.forEach(Hashtable.java:879) [?:1.8.0_181]
        at com.zaxxer.hikari.util.PropertyElf.setTargetFromProperties(PropertyElf.java:52) [HikariCP-2.5.1.jar:?]
        at com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:315) [HikariCP-2.5.1.jar:?]
        at com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:108) [HikariCP-2.5.1.jar:?]
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:99) [HikariCP-2.5.1.jar:?]
        at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:71) [HikariCP-2.5.1.jar:?]
        at net.corda.node.internal.DataSourceFactory.createDataSource(DataSourceFactory.kt:38) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.DataSourceFactory.createDataSource$default(DataSourceFactory.kt:35) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.AbstractNodeKt.configureDatabase(AbstractNode.kt:910) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.AbstractNode.initialiseDatabasePersistence(AbstractNode.kt:680) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.Node.initialiseDatabasePersistence(Node.kt:345) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.AbstractNode.start(AbstractNode.kt:206) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.Node.start(Node.kt:359) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.NodeStartup.startNode(NodeStartup.kt:145) [corda-node-3.2-corda.jar:?]
        at net.corda.node.internal.NodeStartup.run(NodeStartup.kt:115) [corda-node-3.2-corda.jar:?]
        at net.corda.node.Corda.main(Corda.kt:13) [corda-node-3.2-corda.jar:?]
```

Which, after some debugging, turns out to be because the config component treats `dataSource.url` as an object containing `url` 
and the base config specifies this as `"dataSource.url"` (note quotes).

Interestingly, the database initialisation first passes in the postgres url correctly, then promptly ignores it, and tries injecting the default (H2) url to the postgres driver, which fails because it's not a postgres URL :-)

To work around this, this config works:

```
        extraConfig = [
                jarDirs               : ['/Library/Java/postgres'],
                'dataSourceProperties': [
                        'dataSourceClassName': 'org.postgresql.ds.PGSimpleDataSource',
                        '"dataSource.url"'     : 'jdbc:postgresql://localhost:5432/postgres',
                        '"dataSource.user"'    : 'corda',
                        '"dataSource.password"': 'corda'
                ],
                'database'            : [
                        'transactionIsolationLevel': 'READ_COMMITTED',
                        'schema'                   : 'corda'
                ]
        ]
```

which generates:

```
dataSourceProperties {
    "dataSource.password"=corda
    "dataSource.url"="jdbc:postgresql://localhost:5432/postgres"
    "dataSource.user"=corda
    dataSourceClassName="org.postgresql.ds.PGSimpleDataSource"
}
database {
    schema=corda
    transactionIsolationLevel="READ_COMMITTED"
}
jarDirs=[
    "/Library/Java/postgres"
]
```

Which works.

**Comment by mikehearn** [comment]
Filed as https://r3-cev.atlassian.net/browse/CORDA-2080

**Comment by szymonsztuka** [comment]
@dazraf The issue is fixed on master (will be in Corda 4.0) https://r3-cev.atlassian.net/browse/CORDA-1277
We would always have similar problems when any property key with compound name (e.g. dataSource.url) is wrapped in double quotes and overwritten without double quotes or vice versa. To simplify in Corda 4.0 double quotes are forbidden for key values, which solved the issue.

In your example the configs are the same, I presume you meant in the first one keys are without double quotes, e.g.:  'dataSource.url'     : 'jdbc:postgresql://localhost:5432/postgres',

For the  issue - ``schema`` field is ignored - yes, Corda Open Source doesn't use ``schema``, the original answer on StackOverflow must have copy this from  Corda Enterprise. 
You still should be able to setup schema in JDBC connection string  (see https://stackoverflow.com/questions/4168689/is-it-possible-to-specify-the-schema-when-connecting-to-postgres-with-jdbc).

**Comment by szymonsztuka** [comment]
Closing, the issue is resolved in Corda 4.0.

**Comment by cnusrini** [comment]
Hi ALl, can you please share the sample configuration file to configure to postgresql ?
So, I need to include the configurations in individual node's node.config and also in the build.gradle, is that correct? Or are there any other location do I need to update?
